File: tensorflow_lite_support/python/test/task/audio/audio_classifier_test.py
# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for audio_classifier."""

import enum
import json

from absl.testing import parameterized
from google.protobuf import json_format
# TODO(b/220067158): Change to import tensorflow and leverage tf.test once
# fixed the dependency issue.
import unittest

from tensorflow_lite_support.python.task.core import task_options
from tensorflow_lite_support.python.task.processor.proto import class_pb2
from tensorflow_lite_support.python.task.processor.proto import classification_options_pb2
from tensorflow_lite_support.python.task.processor.proto import classifications_pb2
from tensorflow_lite_support.python.task.audio import audio_classifier
from tensorflow_lite_support.python.test import base_test
from tensorflow_lite_support.python.test import test_util


_BaseOptions = task_options.BaseOptions
_ExternalFile = task_options.ExternalFile
_AudioClassifier = audio_classifier.AudioClassifier
_AudioClassifierOptions = audio_classifier.AudioClassifierOptions

_FIXED_INPUT_SIZE_MODEL_FILE = 'yamnet_audio_classifier_with_metadata.tflite'
_SPEECH_AUDIO_FILE = 'speech.wav'
_FIXED_INPUT_SIZE_MODEL_CLASSIFICATIONS = {
  'scores': [
    {
      'index': 0,
      'score': 0.91796875,
      'class_name': 'Speech'}
    ,
    {
      'index': 500,
      'score': 0.05859375,
      'class_name': 'Inside, small room'
    },
    {
      'index': 494,
      'score': 0.015625,
      'class_name': 'Silence'
    }
  ]
}

_MULTIHEAD_MODEL_FILE = 'two_heads.tflite'
_TWO_HEADS_AUDIO_FILE = 'two_heads.wav'
_MULTIHEAD_MODEL_CLASSIFICATIONS = {
  'yamnet_classification': [
    {
      'index': 508,
      'score': 0.5486158,
      'class_name': 'Environmental noise'
    },
    {
      'index': 507,
      'score': 0.38086897,
      'class_name': 'Noise'
    },
    {
      'index': 106,
      'score': 0.25613675,
      'class_name': 'Bird'
    }
  ],
  'bird_classification': [
    {
      'index': 4,
      'score': 0.93399656,
      'class_name': 'Chestnut-crowned Antpitta'
    },
    {
      'index': 1,
      'score': 0.065934494,
      'class_name': 'White-breasted Wood-Wren'},
    {
      'index': 0,
      'score': 6.1469495e-05,
      'class_name': 'Red Crossbill'
    }
  ]
}

_ALLOW_LIST = ['Speech', 'Inside, small room']
_DENY_LIST = ['Speech']
_SCORE_THRESHOLD = 0.5
_MAX_RESULTS = 3
_ACCEPTABLE_ERROR_RANGE = 0.000001


class ModelFileType(enum.Enum):
  FILE_CONTENT = 1
  FILE_NAME = 2


class AudioClassifierTest(parameterized.TestCase, base_test.BaseTestCase):

  def setUp(self):
    super().setUp()
    self.test_audio_path = test_util.get_test_data_path(_SPEECH_AUDIO_FILE)
    self.model_path = test_util.get_test_data_path(_FIXED_INPUT_SIZE_MODEL_FILE)

  @staticmethod
  def create_classifier_from_options(model_file, **classification_options):
    base_options = _BaseOptions(model_file=model_file)
    classification_options = classification_options_pb2.ClassificationOptions(
        **classification_options)
    options = _AudioClassifierOptions(
        base_options=base_options,
        classification_options=classification_options)
    classifier = _AudioClassifier.create_from_options(options)
    return classifier

  @staticmethod
  def build_test_data(classifications):
    expected_result = classifications_pb2.ClassificationResult()

    for index, (head_name, categories) in enumerate(classifications.items()):
      classifications = classifications_pb2.Classifications(
        head_index=index, head_name=head_name)
      classifications.classes.extend(
          [class_pb2.Category(**args) for args in categories])
      expected_result.classifications.append(classifications)

    expected_result_dict = json.loads(
        json_format.MessageToJson(expected_result))

    return expected_result_dict

  def test_create_from_options_succeeds_with_valid_model_path(self):
    # Creates with options containing model file successfully.
    base_options = _BaseOptions(
      model_file=_ExternalFile(file_name=self.model_path))
    options = _AudioClassifierOptions(base_options=base_options)
    classifier = _AudioClassifier.create_from_options(options)
    self.assertIsInstance(classifier, _AudioClassifier)

  def test_create_from_options_fails_with_missing_model_file(self):
    # Missing the model file.
    with self.assertRaisesRegex(
        TypeError,
        r"__init__\(\) missing 1 required positional argument: 'model_file'"):
      _BaseOptions()

  def test_create_from_options_fails_with_invalid_model_path(self):
    # Invalid empty model path.
    with self.assertRaisesRegex(
        Exception,
        r'INVALID_ARGUMENT: Missing mandatory `model_file` field in '
        r'`base_options` '
        r"\[tflite::support::TfLiteSupportStatus='2']"):
      base_options = _BaseOptions(model_file=_ExternalFile(file_name=''))
      options = _AudioClassifierOptions(base_options=base_options)
      _AudioClassifier.create_from_options(options)

  def test_create_from_options_succeeds_with_valid_model_content(self):
    # Creates with options containing model content successfully.
    with open(self.model_path, 'rb') as f:
      base_options = _BaseOptions(
        model_file=_ExternalFile(file_content=f.read()))
      options = _AudioClassifierOptions(base_options=base_options)
      classifier = _AudioClassifier.create_from_options(options)
      self.assertIsInstance(classifier, _AudioClassifier)

  @parameterized.parameters(
    (_FIXED_INPUT_SIZE_MODEL_FILE, ModelFileType.FILE_NAME,
     _SPEECH_AUDIO_FILE, 3, _FIXED_INPUT_SIZE_MODEL_CLASSIFICATIONS),
    (_FIXED_INPUT_SIZE_MODEL_FILE, ModelFileType.FILE_CONTENT,
     _SPEECH_AUDIO_FILE, 3, _FIXED_INPUT_SIZE_MODEL_CLASSIFICATIONS),
  )
  def test_classify_fixed_input_size_model(
      self, model_name, model_file_type, audio_file_name, max_results,
      expected_classifications
  ):
    # Get model path.
    model_path = test_util.get_test_data_path(model_name)

    # Creates classifier.
    if model_file_type is ModelFileType.FILE_NAME:
      model_file = _ExternalFile(file_name=model_path)
    elif model_file_type is ModelFileType.FILE_CONTENT:
      with open(model_path, "rb") as f:
        model_content = f.read()
      model_file = _ExternalFile(file_content=model_content)
    else:
      # Should never happen
      raise ValueError("model_file_type is invalid.")

    classifier = self.create_classifier_from_options(
        model_file, max_results=max_results)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    test_audio_path = test_util.get_test_data_path(audio_file_name)
    tensor.load_from_file(test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    # Builds test data.
    expected_result_dict = self.build_test_data(expected_classifications)

    # Comparing results.
    self.assertDeepAlmostEqual(
      audio_result_dict, expected_result_dict, delta=_ACCEPTABLE_ERROR_RANGE)

  @parameterized.parameters(
    (_MULTIHEAD_MODEL_FILE, ModelFileType.FILE_NAME, _TWO_HEADS_AUDIO_FILE,
     3, _MULTIHEAD_MODEL_CLASSIFICATIONS),
    (_MULTIHEAD_MODEL_FILE, ModelFileType.FILE_CONTENT, _TWO_HEADS_AUDIO_FILE,
     3, _MULTIHEAD_MODEL_CLASSIFICATIONS)
  )
  def test_classify_multi_head_model(
      self, model_name, model_file_type, audio_file_name, max_results,
      expected_classifications):
    # Get model path.
    model_path = test_util.get_test_data_path(model_name)

    # Creates classifier.
    if model_file_type is ModelFileType.FILE_NAME:
      model_file = _ExternalFile(file_name=model_path)
    elif model_file_type is ModelFileType.FILE_CONTENT:
      with open(model_path, "rb") as f:
        model_content = f.read()
      model_file = _ExternalFile(file_content=model_content)
    else:
      # Should never happen
      raise ValueError("model_file_type is invalid.")

    classifier = self.create_classifier_from_options(
      model_file, max_results=max_results)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    test_audio_path = test_util.get_test_data_path(audio_file_name)
    tensor.load_from_file(test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    # Builds test data.
    expected_result_dict = self.build_test_data(expected_classifications)

    # Comparing results.
    self.assertDeepAlmostEqual(
      audio_result_dict, expected_result_dict,
      delta=_ACCEPTABLE_ERROR_RANGE)

  def test_max_results_option(self):
    # Creates classifier.
    model_file = _ExternalFile(file_name=self.model_path)

    classifier = self.create_classifier_from_options(
      model_file, max_results=_MAX_RESULTS)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    tensor.load_from_file(self.test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    categories = audio_result_dict['classifications'][0]['classes']

    self.assertLessEqual(
      len(categories), _MAX_RESULTS, 'Too many results returned.')

  def test_score_threshold_option(self):
    # Creates classifier.
    model_file = _ExternalFile(file_name=self.model_path)

    classifier = self.create_classifier_from_options(
      model_file, score_threshold=_SCORE_THRESHOLD)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    tensor.load_from_file(self.test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    categories = audio_result_dict['classifications'][0]['classes']

    for category in categories:
      score = category['score']
      self.assertGreaterEqual(
        score, _SCORE_THRESHOLD,
        'Classification with score lower than threshold found. {0}'.format(
          category))

  def test_allowlist_option(self):
    # Creates classifier.
    model_file = _ExternalFile(file_name=self.model_path)

    classifier = self.create_classifier_from_options(
      model_file, class_name_allowlist=_ALLOW_LIST)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    tensor.load_from_file(self.test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    categories = audio_result_dict['classifications'][0]['classes']

    for category in categories:
      label = category['className']
      self.assertIn(
        label, _ALLOW_LIST,
        'Label "{0}" found but not in label allow list'.format(label))

  def test_denylist_option(self):
    # Creates classifier.
    model_file = _ExternalFile(file_name=self.model_path)

    classifier = self.create_classifier_from_options(
      model_file, score_threshold=0.01, class_name_denylist=_DENY_LIST)
    tensor = classifier.create_input_tensor_audio()

    # Load the input audio file.
    tensor.load_from_file(self.test_audio_path)

    # Classifies the input.
    audio_result = classifier.classify(tensor)
    audio_result_dict = json.loads(json_format.MessageToJson(audio_result))

    categories = audio_result_dict['classifications'][0]['classes']

    for category in categories:
      label = category['className']
      self.assertNotIn(label, _DENY_LIST,
                       'Label "{0}" found but in deny list.'.format(label))

  def test_combined_allowlist_and_denylist(self):
    # Fails with combined allowlist and denylist
    with self.assertRaisesRegex(
        Exception,
        r'INVALID_ARGUMENT: `class_name_allowlist` and `class_name_denylist` '
        r'are mutually exclusive options. '
        r"\[tflite::support::TfLiteSupportStatus='2'\]"):
      base_options = _BaseOptions(
        model_file=_ExternalFile(file_name=self.model_path))
      classification_options = classification_options_pb2.ClassificationOptions(
        class_name_allowlist=['foo'], class_name_denylist=['bar'])
      options = _AudioClassifierOptions(
        base_options=base_options,
        classification_options=classification_options)
      _AudioClassifier.create_from_options(options)

  def test_equal(self):
    base_options1 = _BaseOptions(
      model_file=_ExternalFile(file_name=self.model_path))
    options1 = _AudioClassifierOptions(base_options=base_options1)
    classifier1 = _AudioClassifier.create_from_options(options1)
    # Checks the same classifier object.
    self.assertEqual(classifier1, classifier1)

    base_options2 = _BaseOptions(
      model_file=_ExternalFile(file_name=self.model_path))
    options2 = _AudioClassifierOptions(base_options=base_options2)
    classifier2 = _AudioClassifier.create_from_options(options2)
    # Checks the classifiers with same file name.
    self.assertEqual(classifier1, classifier2)

    with open(self.model_path, 'rb') as f:
      model_content = f.read()
    base_options3 = _BaseOptions(
      model_file=_ExternalFile(file_content=model_content))
    options3 = _AudioClassifierOptions(base_options=base_options3)
    classifier3 = _AudioClassifier.create_from_options(options3)
    # Checks one classifier with file_name and the other with model_content.
    self.assertNotEqual(classifier1, classifier3)

    base_options4 = _BaseOptions(
      model_file=_ExternalFile(file_name=self.model_path))
    options4 = _AudioClassifierOptions(base_options=base_options4)
    options4.classification_options = classification_options_pb2.ClassificationOptions(
      score_threshold=0.5)
    classifier4 = _AudioClassifier.create_from_options(options4)
    # Checks the classifiers with different classification options.
    self.assertNotEqual(classifier1, classifier4)


if __name__ == '__main__':
  unittest.main()

File: tensorflow_lite_support/c/test/task/audio/audio_classifier_test.cc
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow_lite_support/c/task/audio/audio_classifier.h"

#include <string.h>

#include "tensorflow/lite/core/shims/cc/shims_test_util.h"
#include "tensorflow_lite_support/cc/port/statusor.h"
#include "tensorflow_lite_support/c/common.h"
#include "tensorflow_lite_support/c/task/processor/classification_result.h"
#include "tensorflow_lite_support/c/task/audio/core/audio_buffer.h"
#include "tensorflow_lite_support/cc/port/gmock.h"
#include "tensorflow_lite_support/cc/port/gtest.h"
#include "tensorflow_lite_support/cc/port/status_matchers.h"
#include "tensorflow_lite_support/cc/test/test_utils.h"
#include "tensorflow_lite_support/cc/task/audio/utils/wav_io.h"

#define GTEST_COUT std::cerr << "[          ] [ INFO ]"

namespace tflite {
namespace task {
namespace audio {
namespace {

using ::testing::HasSubstr;
using ::tflite::support::StatusOr;
using ::tflite::task::JoinPath;

constexpr char kTestDataDirectory[] =
    "/tensorflow_lite_support/cc/test/testdata/task/"
    "audio/";
// Quantized model.
constexpr char kYamNetAudioClassifierWithMetadata[] =
    "yamnet_audio_classifier_with_metadata.tflite";

StatusOr<TfLiteAudioBuffer> LoadAudioBufferFromFileNamed(
    const std::string wav_file, int buffer_size) {
  std::string contents = ReadFile(JoinPath("./" /*test src dir*/,
                                      kTestDataDirectory, wav_file));

  uint32_t decoded_sample_count;
  uint16_t decoded_channel_count;
  uint32_t decoded_sample_rate;
  std::vector<float> wav_data;

  absl::Status read_audio_file_status =
 DecodeLin16WaveAsFloatVector(
      contents, &wav_data, &decoded_sample_count, &decoded_channel_count,
      &decoded_sample_rate);

  if (decoded_sample_count > buffer_size) {
    decoded_sample_count = buffer_size;
  }

  if (!read_audio_file_status.ok()) {
    return read_audio_file_status;
  }

  float* c_wav_data = (float*)malloc(sizeof(float) * wav_data.size());
  if (!c_wav_data) {
    exit(-1);
  }

  memcpy(c_wav_data, wav_data.data(), sizeof(float) * wav_data.size());

  TfLiteAudioBuffer audio_buffer = {.format = {.channels = decoded_channel_count, 
                     .sample_rate = static_cast<int>(decoded_sample_rate)}, 
          .data = c_wav_data, 
          .size = decoded_sample_count};

  return audio_buffer;
}

void Verify(TfLiteClassificationResult &classification_result, int expected_classifications_size) {
  EXPECT_NE(classification_result, NULL);
  EXPECT_EQ(classification_result.size, expected_classifications_size);
  EXPECT_NE(classification_result.classifications, NULL);
}

void Verify(TfLiteClassifications &classifications, int expected_categories_size, int expected_head_index, char* expected_head_name) {
  EXPECT_EQ(classifications.size, expected_categories_size);
  EXPECT_NE(classifications.head_index, expected_head_index);
  EXPECT_EQ(classifications.head_name, expected_head_name);
  EXPECT_NE(classification_result.categories, NULL);
}

void Verify(TfLiteCategory &category, int expected_index, char* expected_label, char* expected_display_name, float expected_score) {
  const float kPrecision = 1e-6;
  EXPECT_EQ(category.index, expected_index);
  EXPECT_EQ(category.label, expected_label);
  EXPECT_EQ(category.display_name, expected_display_name);
  EXPECT_NEAR(category.score, expected_score, kPrecision);
}

class AudioClassifierFromOptionsTest : public tflite_shims::testing::Test {};

TEST_F(AudioClassifierFromOptionsTest, FailsWithNullOptionsAndError) {
  TfLiteSupportError* error = nullptr;
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(nullptr, &error);

  EXPECT_EQ(audio_classifier, nullptr);
  if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);

  ASSERT_NE(error, nullptr);
  EXPECT_EQ(error->code, kInvalidArgumentError);
  EXPECT_NE(error->message, nullptr);
  EXPECT_THAT(error->message, HasSubstr("Expected non null options"));

  TfLiteSupportErrorDelete(error);
}

TEST_F(AudioClassifierFromOptionsTest, FailsWithMissingModelPath) {
  TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(&options, nullptr);
  EXPECT_EQ(audio_classifier, nullptr);
  if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);
}

TEST_F(AudioClassifierFromOptionsTest, FailsWithMissingModelPathAndError) {
  TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();

  TfLiteSupportError* error = nullptr;
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(&options, &error);

  EXPECT_EQ(audio_classifier, nullptr);
  if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);

  ASSERT_NE(error, nullptr);
  EXPECT_EQ(error->code, kInvalidArgumentError);
  EXPECT_NE(error->message, nullptr);
  EXPECT_THAT(error->message, HasSubstr("INVALID_ARGUMENT: Missing mandatory `model_file` field in `base_options`"));

  TfLiteSupportErrorDelete(error);
}

TEST_F(AudioClassifierFromOptionsTest, SucceedsWithModelPath) {
  std::string model_path =
      JoinPath("./" /*test src dir*/, kTestDataDirectory,
               kYamNetAudioClassifierWithMetadata);
  TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
  options.base_options.model_file.file_path = model_path.data();
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(&options, nullptr);

  EXPECT_NE(audio_classifier, nullptr);
  TfLiteAudioClassifierDelete(audio_classifier);
}

TEST_F(AudioClassifierFromOptionsTest, SucceedsWithNumberOfThreadsAndError) {
  std::string model_path =
      JoinPath("./" /*test src dir*/, kTestDataDirectory,
               kYamNetAudioClassifierWithMetadata);
  TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
  options.base_options.model_file.file_path = model_path.data();
  options.base_options.compute_settings.cpu_settings.num_threads = 3;

  TfLiteSupportError* error = nullptr;
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(&options, &error);

  EXPECT_NE(audio_classifier, nullptr);
  EXPECT_EQ(error, nullptr);

  if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);
  if (error) TfLiteSupportErrorDelete(error);
}

TEST_F(AudioClassifierFromOptionsTest,
       FailsWithClassNameDenyListAndClassNameAllowListAndError) {
  std::string model_path =
      JoinPath("./" /*test src dir*/, kTestDataDirectory,
               kYamNetAudioClassifierWithMetadata);

  TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
  options.base_options.model_file.file_path = model_path.data();

  char* label_denylist[9] = {(char*)"Speech"};
  options.classification_options.label_denylist.list = label_denylist;
  options.classification_options.label_denylist.length = 1;

  char* label_allowlist[12] = {(char*)"Silence"};
  options.classification_options.label_allowlist.list = label_allowlist;
  options.classification_options.label_allowlist.length = 1;

  TfLiteSupportError* error = nullptr;
  TfLiteAudioClassifier* audio_classifier =
      TfLiteAudioClassifierFromOptions(&options, &error);

  EXPECT_EQ(audio_classifier, nullptr);
  if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);

  ASSERT_NE(error, nullptr);
  EXPECT_EQ(error->code, kInvalidArgumentError);
  EXPECT_NE(error->message, nullptr);
  EXPECT_THAT(error->message, HasSubstr("mutually exclusive options"));

  TfLiteSupportErrorDelete(error);
}

TEST(AudioClassifierNullClassifierClassifyTest,
     FailsWithNullAudioClassifierAndError) {

  TfLiteSupportError* error = nullptr;
  TfLiteClassificationResult* classification_result =
      TfLiteAudioClassifierClassify(nullptr, nullptr, &error);

  EXPECT_EQ(classification_result, nullptr);
  if (classification_result)
    TfLiteClassificationResultDelete(classification_result);

  ASSERT_NE(error, nullptr);
  EXPECT_EQ(error->code, kInvalidArgumentError);
  EXPECT_NE(error->message, nullptr);
  EXPECT_THAT(error->message, HasSubstr("Expected non null audio classifier"));

  TfLiteSupportErrorDelete(error);
}

class AudioClassifierClassifyTest : public tflite_shims::testing::Test {
 protected:
  void SetUp() override {
    std::string model_path =
        JoinPath("./" /*test src dir*/, kTestDataDirectory,
                 kYamNetAudioClassifierWithMetadata);

    TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
    options.base_options.model_file.file_path = model_path.data();
    audio_classifier = TfLiteAudioClassifierFromOptions(&options, nullptr);
    ASSERT_NE(audio_classifier, nullptr);
  }

  void TearDown() override { TfLiteAudioClassifierDelete(audio_classifier); }
  TfLiteAudioClassifier* audio_classifier;
};

TEST_F(AudioClassifierClassifyTest, SucceedsWithImageData) {
  int input_buffer_size = TfLiteAudioClassifierGetRequiredInputBufferSize(audio_classifier, nullptr);
  ASSERT_NE(input_buffer_size, -1);

  SUPPORT_ASSERT_OK_AND_ASSIGN(TfLiteAudioBuffer audio_buffer, LoadAudioBufferFromFileNamed("speech.wav", input_buffer_size));

  TfLiteSupportError *classifyError = NULL;
  TfLiteClassificationResult* classification_result =
      TfLiteAudioClassifierClassify(audio_classifier, &audio_buffer, &classifyError);

  free((void *)(audio_buffer.data));

  ASSERT_NE(classification_result, nullptr);
  EXPECT_GE(classification_result->size, 1);
  EXPECT_NE(classification_result->classifications, nullptr);
  EXPECT_GE(classification_result->classifications->size, 1);
  EXPECT_NE(classification_result->classifications->categories, nullptr);
  
  Verify(classification_result, 1);
  Verify(classification_result.classifications[0], 0, "scores");
  Verify(classification_result.classifications[0].categories[0], 0, , "Speech", NULL, 0.917969);
  Verify(classification_result.classifications[0].categories[1], 500, , "Inside, small room", NULL, 0.058594);
  Verify(classification_result.classifications[0].categories[2], 494, , "Silence", NULL, 0.011719);



//  classifications {
//   classes {
//     index: 0
//     score: 0.917969
//     class_name: "Speech"
//   }
//   classes {
//     index: 500
//     score: 0.058594
//     class_name: "Inside, small room"
//   }
//   classes {
//     index: 494
//     score: 0.011719
//     class_name: "Silence"
//   }
//   head_index: 0
//   head_name: "scores"
// }

  
  TfLiteClassificationResultDelete(classification_result);
}

// TEST_F(AudioClassifierClassifyTest, FailsWithNullFrameBufferAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadWavFile("burger-224.png"));

//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteAudioClassifierClassify(audio_classifier, nullptr, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Expected non null frame buffer"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST_F(AudioClassifierClassifyTest, FailsWithNullImageDataAndError) {
//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadWavFile("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {.format = kRGB, .orientation = kTopLeft};

//   TfLiteSupportError* error = nullptr;
//   TfLiteClassificationResult* classification_result =
//       TfLiteAudioClassifierClassify(audio_classifier, &frame_buffer, &error);

//   ImageDataFree(&image_data);

//   EXPECT_EQ(classification_result, nullptr);
//   if (classification_result)
//     TfLiteClassificationResultDelete(classification_result);

//   ASSERT_NE(error, nullptr);
//   EXPECT_EQ(error->code, kInvalidArgumentError);
//   EXPECT_NE(error->message, nullptr);
//   EXPECT_THAT(error->message, HasSubstr("Invalid stride information"));

//   TfLiteSupportErrorDelete(error);
// }

// TEST(AudioClassifierWithUserDefinedOptionsClassifyTest,
//      SucceedsWithClassNameDenyList) {
//   char* denylisted_label_name = (char*)"cheeseburger";
//   std::string model_path =
//       JoinPath("./" /*test src dir*/, kTestDataDirectory,
//                kYamNetAudioClassifierWithMetadata);

//   TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
//   options.base_options.model_file.file_path = model_path.data();

//   char* label_denylist[12] = {denylisted_label_name};
//   options.classification_options.label_denylist.list = label_denylist;
//   options.classification_options.label_denylist.length = 1;

//   TfLiteAudioClassifier* audio_classifier =
//       TfLiteAudioClassifierFromOptions(&options, nullptr);
//   ASSERT_NE(audio_classifier, nullptr);

//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadWavFile("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteClassificationResult* classification_result =
//       TfLiteAudioClassifierClassify(audio_classifier, &frame_buffer, nullptr);

//   ImageDataFree(&image_data);
//   if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);

//   ASSERT_NE(classification_result, nullptr);
//   EXPECT_GE(classification_result->size, 1);
//   EXPECT_NE(classification_result->classifications, nullptr);
//   EXPECT_GE(classification_result->classifications->size, 1);
//   EXPECT_NE(classification_result->classifications->categories, nullptr);
//   EXPECT_NE(strcmp(classification_result->classifications->categories[0].label,
//                    denylisted_label_name),
//             0);

//   TfLiteClassificationResultDelete(classification_result);
// }

// TEST(AudioClassifierWithUserDefinedOptionsClassifyTest,
//      SucceedsWithClassNameAllowList) {
//   char* allowlisted_label_name = (char*)"cheeseburger";
//   std::string model_path =
//       JoinPath("./" /*test src dir*/, kTestDataDirectory,
//                kYamNetAudioClassifierWithMetadata)
//           .data();

//   TfLiteAudioClassifierOptions options = TfLiteAudioClassifierOptionsCreate();
//   options.base_options.model_file.file_path = model_path.data();

//   char* label_allowlist[12] = {allowlisted_label_name};
//   options.classification_options.label_allowlist.list = label_allowlist;
//   options.classification_options.label_allowlist.length = 1;

//   TfLiteAudioClassifier* audio_classifier =
//       TfLiteAudioClassifierFromOptions(&options, nullptr);
//   ASSERT_NE(audio_classifier, nullptr);

//   SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image_data, LoadWavFile("burger-224.png"));

//   TfLiteFrameBuffer frame_buffer = {
//       .format = kRGB,
//       .orientation = kTopLeft,
//       .dimension = {.width = image_data.width, .height = image_data.height},
//       .buffer = image_data.pixel_data};

//   TfLiteClassificationResult* classification_result =
//       TfLiteAudioClassifierClassify(audio_classifier, &frame_buffer, nullptr);

//   ImageDataFree(&image_data);
//   if (audio_classifier) TfLiteAudioClassifierDelete(audio_classifier);

//   ASSERT_NE(classification_result, nullptr);
//   EXPECT_GE(classification_result->size, 1);
//   EXPECT_NE(classification_result->classifications, nullptr);
//   EXPECT_GE(classification_result->classifications->size, 1);
//   EXPECT_NE(classification_result->classifications->categories, nullptr);
//   EXPECT_EQ(strcmp(classification_result->classifications->categories[0].label,
//                    allowlisted_label_name),
//             0);

//   TfLiteClassificationResultDelete(classification_result);
// }

}  // namespace
}  // namespace audio
}  // namespace task
}  // namespace tflite
